2021-Firmware-Bootcamp/Core/Src/main.c

```diff
@@ -0,0 +1,221 @@
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  if (HAL_TIM_Base_Start(&htim1) != HAL_OK) // Start timer
+  {
+	  Error_Handler();
+  }
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET); // Disable communication with ADC
+  while (1)
+  {
+    /* USER CODE END WHILE */
+
+    /* USER CODE BEGIN 3 */
+	  uint8_t tx_data[3];
+	  uint8_t rx_data[3];
+
+	  tx_data[0] = 0b00000001; // Alignment bit x7, Start bit
+	  tx_data[1] = 0b10000000; // Single mode bit, Channel config bit *3, Don't care bit *4
+	  // tx_data[2] is only don't care bits
+
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET); // Initiate communication with ADC
+
+	  if (HAL_SPI_TransmitReceive(&hspi1, tx_data, rx_data, 3, HAL_MAX_DELAY) != HAL_OK)
+	  {
+		  Error_Handler();
+	  }
+
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET); // Disable communication with ADC
+
+	  uint16_t pot_value = (rx_data[1] << 8) | rx_data[0];
+	  pot_value &= 0b1111111111; // Take only last 10 bits
+
+	  const uint16_t FIVE_PERCENT_COUNTER_PERIOD = 3200; // 64000 * 0.05 = 3200
+	  const uint16_t POT_MAX_VALUE = 1023; // Maximum 10 bit value
+	  uint16_t duty_cycle = FIVE_PERCENT_COUNTER_PERIOD * (1 + pot_value / POT_MAX_VALUE); // I'm not exactly sure if the math checks out
+
+	  // Set compare register to the calculated "on counts" number
+	  __HAL_TIM_SET_COMPARE(&htim1, TIM_CHANNEL_1, duty_cycle);
+
+	  HAL_Delay(10);
+  }
+  if (HAL_TIM_Base_Stop(&htim1) != HAL_OK) // Stop timer (not sure if this is or not necessary as it is out of while loop)
+  {
+  	  Error_Handler();
+  }
```

Comment:
This should not be here, if you are stopping the clock, then you would also be stopping all the functions related to this clock (our PWM signal) which is not good.